apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Trace/GrpcTraceHeaderFactory.cs

```diff
@@ -0,0 +1,170 @@
+
+        private const byte TraceOptionsFieldId = 2;
+        private static readonly byte[] TraceOptions = new byte[] { 1 };
+
+        private const string MetadataEntryKey = "grpc-trace-bin";
+
+        private readonly Func<IManagedTracer> _managedTracerFactory;
+
+        /// <summary>
+        /// Creates a new instance using the given delegate to obtain the "current" tracer.
+        /// </summary>
+        /// <param name="managedTracerFactory">A delegate used to obtain the "current" tracer
+        /// for each request. The delegate should therefore be thread-safe.</param>
+        public GrpcTraceHeaderFactory(Func<IManagedTracer> managedTracerFactory)
+        {
+            _managedTracerFactory = GaxPreconditions.CheckNotNull(managedTracerFactory, nameof(managedTracerFactory));
+        }
+
+        /// <summary>
+        /// Starts a <see cref="ISpan"/> with the correct name and kind for gRPC trace propagation.
+        /// The span name will be of the form <code>packageName.serviceName/methodName</code>
+        /// and the span kind will be of the form <see cref="SpanKind.RpcClient"/>.
+        /// This method should be called before any gRPC call that will contain trace propagation
+        /// headers.
+        /// </summary>
+        /// <param name="packageName">The name of the gRPC package.</param>
+        /// <param name="serviceName">The name of the gRPC service.</param>
+        /// <param name="methodName">The name of the gRPC method</param>
+        /// <returns>An <see cref="ISpan"/> appropiate for gRPC trace propagation.</returns>
+        public ISpan GetSpanForPropagation(string packageName, string serviceName, string methodName)
+        {
+            var tracer = _managedTracerFactory();
+            return tracer.StartSpan(
+                $"{packageName}.{serviceName}/{methodName}", StartSpanOptions.Create(SpanKind.RpcClient));
+        }
+
+        /// <summary>
+        /// Builds the <see cref="CallSettings"/> to use in a gRPC call so that
+        /// trace is propagated.
+        /// <see cref="GetSpanForPropagation(string, string, string)"/> needs to be called
+        /// before the gRPC call that used these <see cref="CallSettings"/>.
+        /// </summary>
+        public CallSettings GetCallSettingsForPropagation()
+        {
+            return CallSettings.FromHeaderMutation(GetHeaderMutationForPropagation());
+        }
+
+        /// <summary>
+        /// Returns the header mutation action that will change gRPC headers so
+        /// that trace is propagated to the gRPC request.
+        /// </summary>
+        public Action<Metadata> GetHeaderMutationForPropagation()
+        {
+            return metadata => metadata.Add(GetGrpcHeader());
+        }
+
+        /// <summary>
+        /// Builds a <see cref="Metadata.Entry"/>, that is a gRPC header, for trace propagation.
+        /// The header is built based on the tracer returned by the manageTracerFactory with
```

Comment:
Nit: given that ManagedTracerFactory isn't a type, I'd just expand it to "managed tracer factory".

Reply:
Done